Sources/StreamChat/Workers/Background/ConnectionRecoveryHandler.swift

```diff
@@ -0,0 +1,257 @@
+            onEnteringBackground: { [weak self] in self?.appDidEnterBackground() },
+            onEnteringForeground: { [weak self] in self?.appDidBecomeActive() }
+        )
+        
+        internetConnection.notificationCenter.addObserver(
+            self,
+            selector: #selector(internetConnectionAvailabilityDidChange(_:)),
+            name: .internetConnectionAvailabilityDidChange,
+            object: nil
+        )
+    }
+    
+    func unsubscribeFromNotifications() {
+        backgroundTaskScheduler?.stopListeningForAppStateUpdates()
+        
+        internetConnection.notificationCenter.removeObserver(
+            self,
+            name: .internetConnectionStatusDidChange,
+            object: nil
+        )
+    }
+}
+
+// MARK: - Event handlers
+
+extension DefaultConnectionRecoveryHandler {
+    private func appDidBecomeActive() {
+        log.debug("App -> ✅", subsystems: .webSocket)
+        
+        backgroundTaskScheduler?.endTask()
+        
+        reconnectIfNeeded()
+    }
+    
+    private func appDidEnterBackground() {
+        log.debug("App -> 💤", subsystems: .webSocket)
+        
+        guard canBeDisconnected else {
+            // Client is not trying to connect nor connected
+            return
+        }
+        
+        guard keepConnectionAliveInBackground else {
+            // We immediately disconnect
+            disconnectIfNeeded()
+            return
+        }
+        
+        guard let scheduler = backgroundTaskScheduler else { return }
+                
+        let succeed = scheduler.beginTask { [weak self] in
+            log.debug("Background task -> ❌", subsystems: .webSocket)
+            
+            self?.disconnectIfNeeded()
+        }
+        
+        if succeed {
+            log.debug("Background task -> ✅", subsystems: .webSocket)
+        } else {
+            // Can't initiate a background task, close the connection
+            disconnectIfNeeded()
+        }
+    }
+    
+    @objc private func internetConnectionAvailabilityDidChange(_ notification: Notification) {
+        guard let isAvailable = notification.internetConnectionStatus?.isAvailable else { return }
+        
+        log.debug("Internet -> \(isAvailable ? "✅" : "❌")", subsystems: .webSocket)
+        
+        if isAvailable {
+            reconnectIfNeeded()
+        } else {
+            disconnectIfNeeded()
+        }
+    }
+    
+    func webSocketClient(_ client: WebSocketClient, didUpdateConnectionState state: WebSocketConnectionState) {
+        log.debug("Connection state: \(state)", subsystems: .webSocket)
+        
+        switch state {
+        case .connecting:
+            cancelReconnectionTimer()
+            
+        case .connected:
+            reconnectionStrategy.resetConsecutiveFailures()
+            syncRepository.syncLocalState {
+                log.info("Local state sync completed", subsystems: .offlineSupport)
+            }
```

Comment:
This part needs to be improved and should be planned for offline support to be 100% reliable. 

The only indicator about the connectivity is the `.connected` status. At the same time, we fire another `async` operation to synchronize local state with the remote.

This creates the "buggy" window, in which the integrator might start mutating local state while it's being synced with the remote.

Since the synchronized state is not given to us at the `connected` state, I'd suggest to treat the "connecteded" state that the integrator can listen to as the "websocket.connected && local.state synchronised"

Reply:
While you are right on one side, this has massive implications. Also, bear in mind that the thing that we would end up doing is we would be queuing DB actions (because API calls are already queued, and those undo DB changes if the request then fails). In the case of DB actions, it does not really matter that much when those are done, because they are all going to get through when the queue gets executed, there is no backend that can tell you what can/needs to be done when launching them.
Also, if you you were to do that, we would need to block the whole UI from any kind of actions. This is something that would need a lot of changes in the LLC as well.

All in all, I completely get your point, but TBH I don't see it a major issue (Maybe we discover it is, don't get me wrong, but it doesn't seem to be the case now)

Reply:
imo this needs follow-up. It's behaviour undefined and should be addressed to strengthen LLC Foundation.